### ISSUE TITLE
no-unused-vars incorrect results

### ISSUE THREAD

**Issue by armano2** [bradzacher/eslint-plugin-typescript#179]
**Repro**

```JSON
{
  "rules": {
    "no-unused-vars": "error",
    "typescript/no-unused-vars": "error"
  }
}
```

```TS
type Foo = number

function test<Foo>(bar: Foo) {
    return bar
}
```

**Expected Result**
error that Foo type is not defined

**Actual Result**
all is good

**Additional Info**

There is few issues there:
- markVariableAsUsed is iterating over all variables even if found first one,
- type variables are not validated

** Versions **
| package                    | version |
| -------------------------- | ------- |
| `eslint-plugin-typescript` | `0.13.0` |
| `typescript-eslint-parser` | `4.13.1` |
| `typescript`               | `17.0.1` |

**Comment by bradzacher** [comment]
Digging into this, I don't know what we can do about this without significant work.

Right now our rule sits on top of the base rule, so I think we would have to clone the base rule's code in order to handle this case properly.

This would be a good candidate for using the TS parser services *if* it reports unused vars.

This should be caught by something like `no-redeclare`. Though it won't because we haven't added ts-specific handling to that rule.

**Comment by armano2** [comment]
@bradzacher right now there is no real solution for that. It's require some work on parser before we can start.
I started doing PR with support for additional data in scope for typescript "types" but its not near finish.
When it will be done, we can start thinking about solving this. 

I made this issue for now just to inform ppl that we are aware about it and we will work on it.
